types/number.go

```diff
@@ -0,0 +1,71 @@
+package types
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/shopspring/decimal"
+)
+
+var errInvalidNumberValue = errors.New("invalid value for Number type")
+
+// Number represents a floating point number, and implements json.Unmarshaller and json.Marshaller
+type Number float64
+
+// UnmarshalJSON implements json.Unmarshaler
+func (f *Number) UnmarshalJSON(data []byte) error {
+	switch c := data[0]; c { // From json.decode literalInterface
+	case 'n', 't', 'f': // null, true, false
+		return fmt.Errorf("%w: %s", errInvalidNumberValue, data)
+	case '"': // string
+		if len(data) < 2 || data[len(data)-1] != '"' {
+			return fmt.Errorf("%w: %s", errInvalidNumberValue, data)
+		}
+		data = data[1 : len(data)-1] // Naive Unquote
+	default: // Should be a number
+		if c != '-' && (c < '0' || c > '9') { // Invalid json syntax
+			return fmt.Errorf("%w: %s", errInvalidNumberValue, data)
+		}
+	}
+
+	if len(data) == 0 {
+		*f = Number(0)
+		return nil
+	}
+
+	val, err := strconv.ParseFloat(string(data), 64)
+	if err != nil {
+		return fmt.Errorf("%w: %s", errInvalidNumberValue, data) // We don't use err; We know it's not valid and errInvalidNumberValue is clearer
+	}
+
+	*f = Number(val)
+
+	return nil
+}
+
+// MarshalJSON implements json.Marshaler
```

Comment:
Could you please add here saying that it marshalls as a string? Its the only thing the `convert.StringToFloat` kinda sorta implied, but this loses. One might use a `type.Number` not realising its marshalled into `"1337"`

I'd also accept it being added  to the type declaration instead

Reply:
```
// MarshalJSON implements json.Marshaler by formatting to a json string
// 1337.37 will marshal to "1337.37"
// 0 will marshal to an empty string: ""
```
Fixed a7bd1dff